freemarker/src/test/java/org/jdbi/v3/sqlobject/BindBeanListTest.java

```diff
@@ -0,0 +1,205 @@
+    public interface SomethingWithExplicitAttributeName {
+        @SqlQuery("select id, name from something where (id, name) in (${keys})")
+        List<Something> get(@BindBeanList(value = "keys", propertyNames = {"id", "name"})
+                                    SomethingKey... blarg);
+    }
+
+    //
+
+    @Test
+    public void testSomethingByVarargsWithVarargs() {
+        final SomethingByVarargs s = handle.attach(SomethingByVarargs.class);
+
+        final List<Something> out = s.get(
+                new SomethingKey(1, "1"),
+                new SomethingKey(2, "2"));
+
+        assertThat(out).hasSameElementsAs(expectedSomethings);
+    }
+
+    @Test
+    public void testSomethingByVarargsWithEmptyVarargs() {
+        final SomethingByVarargs s = handle.attach(SomethingByVarargs.class);
+
+        assertThatThrownBy(s::get).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @UseFreemarkerEngine
+    public interface SomethingByVarargs {
+        @SqlQuery("select id, name from something where (id, name) in (${keys})")
+        List<Something> get(@BindBeanList(propertyNames = {"id", "name"}) SomethingKey... keys);
+    }
+
+    //
+
+    @Test
+    public void testSomethingByArrayWithNull() {
+        final SomethingByArray s = handle.attach(SomethingByArray.class);
+
+        assertThatThrownBy(() -> s.get(null)).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void testSomethingByArrayWithEmptyArray() {
+        final SomethingByArray s = handle.attach(SomethingByArray.class);
+
+        assertThatThrownBy(() -> s.get(new SomethingKey[]{})).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void testSomethingByArrayWithNonEmptyArray() {
+        final SomethingByVarargs s = handle.attach(SomethingByVarargs.class);
+
+        final List<Something> out = s.get(
+                new SomethingKey(1, "1"),
+                new SomethingKey(2, "2"));
+
+        assertThat(out).hasSameElementsAs(expectedSomethings);
+    }
+
+    @UseFreemarkerEngine
+    private interface SomethingByArray {
+        @SqlQuery("select id, name from something where (id, name) in (<keys>)")
+        List<Something> get(@BindBeanList(propertyNames = {"id", "name"}) SomethingKey[] keys);
+    }
+
+    //
+
+    @Test
+    public void testSomethingByIterableWithIterable() {
+        final SomethingByIterable s = handle.attach(SomethingByIterable.class);
+
+        final List<Something> out = s.get(() -> Arrays.asList(new SomethingKey(1, "1"),
+                new SomethingKey(2, "2"))
+                .iterator());
+
+        assertThat(out).hasSameElementsAs(expectedSomethings);
+    }
+
+    @Test
+    public void testSomethingByIterableWithEmptyIterable() {
+        final SomethingByIterable s = handle.attach(SomethingByIterable.class);
+
+        assertThatThrownBy(() -> s.get(new ArrayList<>())).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @UseFreemarkerEngine
```

Comment:
Does this annotation want attributes for the variable markers? E.g. `$`, `{`, and `}`

Reply:
I think the best option for now is to get the module battle tested. I can add a pull request on this in the future if needed? I will never need it. Standard freemarker variable markers is good enough ;)

Reply:
Nevermind this one, I was conflating this PR with the one for StringSubstitutor template engine, which allows you to swap out the substitution markers.